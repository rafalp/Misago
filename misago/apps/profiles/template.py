from datetime import timedelta
from django.conf import settings
from django.template import RequestContext as DjangoRequestContext
from django.utils import timezone
from django.utils.importlib import import_module
from misago.models import User

def RequestContext(request, context=None):
    if not context:
        context = {}
    context['fallback'] = request.path
        
    # Find out if we ignore or follow this user
    context['follows'] = False
    context['ignores'] = False
    if request.user.is_authenticated() and request.user.pk != context['profile'].pk:
        context['follows'] = request.user.is_following(context['profile'])
        context['ignores'] = request.user.is_ignoring(context['profile'])
    
    # Find out if this user allows us to see his activity
    if request.user.pk != context['profile'].pk:
        if context['profile'].hide_activity == 2:
            context['hidden'] = True
        if context['profile'].hide_activity == 1:
            context['hidden'] = context['profile'].is_following(request.user)
    else:
        context['hidden'] = False

    # Find out if this user is online:
    if request.user.pk != context['profile'].pk:
        try:
            context['online'] = context['profile'].sessions.filter(admin=False).filter(last__gt=(timezone.now() - timedelta(minutes=10))).order_by('-last')[0:1][0]
        except IndexError:
            context['online'] = False
    else:
        # Fake "right now" time
        context['online'] = {'last': timezone.now()}

    # Sync member
    if context['profile'].sync_profile():
        context['profile'].save(force_update=True)

    context['tabs'] = []
    for extension in settings.PROFILE_EXTENSIONS:
        profile_module = import_module(extension + '.profile')
        try:
            append_links = profile_module.register_profile_extension(request)
            if append_links:
                for link in append_links:
                    link = list(link)
                    token = link[0][link[0].find('_') + 1:]
                    context['tabs'].append({
                                            'route': link[0],
                                            'active': context['tab'] == token,
                                            'name': link[1],
                                            })
        except AttributeError:
            pass
    return DjangoRequestContext(request, context)
