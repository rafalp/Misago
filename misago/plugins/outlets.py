from functools import wraps
from typing import Dict, List, Protocol

from django.http import HttpRequest
from django.template import Context
from django.utils.safestring import SafeString, mark_safe

from .enums import PluginOutlet
from .hooks import ActionHook


class PluginOutletHookAction(Protocol):
    def __call__(
        self, request: HttpRequest, context: dict | Context
    ) -> str | SafeString | None:
        pass


class PluginOutletHook(ActionHook[PluginOutletHookAction]):
    __slots__ = ActionHook.__slots__

    def __call__(
        self, request: HttpRequest, context: dict | Context
    ) -> List[str | SafeString | None]:
        return super().__call__(request, context)


template_outlets: Dict[str, PluginOutletHook] = {}


def create_new_outlet(outlet_name: str):
    if outlet_name in template_outlets:
        raise ValueError(f"Template outlet '{outlet_name}' already exists.")

    template_outlets[plugin_outlet.name] = PluginOutletHook()


for plugin_outlet in PluginOutlet:
    create_new_outlet(plugin_outlet.name)


def append_outlet_action(
    outlet_name: str | PluginOutlet, action: PluginOutletHookAction
):
    get_outlet(outlet_name).append_action(action)


def prepend_outlet_action(
    outlet_name: str | PluginOutlet, action: PluginOutletHookAction
):
    get_outlet(outlet_name).prepend_action(action)


def get_outlet(outlet_name: str | PluginOutlet) -> PluginOutletHook:
    try:
        if isinstance(outlet_name, PluginOutlet):
            return template_outlets[outlet_name.name]
        return template_outlets[outlet_name]
    except KeyError as exc:
        raise KeyError(f"Unknown template outlet: {outlet_name}") from exc


def template_outlet_action(f):
    """Decorator for an outlet action that renders a template with returned context."""

    @wraps(f)
    @mark_safe
    def wrapped_outlet_action(request: HttpRequest, context: Context):
        template_data = f(request, context)
        if template_data is None:
            return ""

        if isinstance(template_data, str):
            return _include_template(template_data, context)

        if isinstance(template_data, tuple):
            template_name, new_context = template_data
            return _include_template(template_name, context, new_context)

        return ""

    return wrapped_outlet_action


def _include_template(
    template_name: str, context: Context, new_context: dict | None = None
):
    """Subset of Django include template tag.

    Works only with Django template engine.
    """
    cache = context.render_context.dicts[0].setdefault("_template_outlet_action", {})
    template = cache.get(template_name)
    if template is None:
        template = context.template.engine.get_template(template_name)
        cache[template_name] = template

    if new_context:
        with context.push(**new_context):
            return template.render(context)

    return template.render(context)
