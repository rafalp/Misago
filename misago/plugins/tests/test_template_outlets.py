from contextlib import contextmanager

from django.template import Context, Template

from ..outlets import (
    PluginOutletName,
    PluginOutletHook,
    append_template_plugin,
    prepend_template_plugin,
    template_outlets,
)


def strong_action(context):
    body = context.get("value") or "none"
    return f"<strong>{body}</strong>"


def em_action(context):
    body = context.get("value") or "none"
    return f"<em>{body}</em>"


def render_template(context: dict | None = None):
    template = Template(
        """
        {% load misago_plugins %}
        <div>{% pluginoutlet "TEST" %}</div>
        """
    )

    return template.render(Context(context or {})).strip()


@contextmanager
def patch_outlets():
    try:
        org_outlets = template_outlets.copy()
        for key in template_outlets:
            template_outlets[key] = PluginOutletHook()
        yield template_outlets
    finally:
        for key, hook in org_outlets.items():
            template_outlets[key] = hook


def test_empty_outlet_renders_nothing(snapshot):
    with patch_outlets():
        html = render_template()

    assert snapshot == html


def test_outlet_renders_appended_plugin(snapshot):
    with patch_outlets():
        append_template_plugin(PluginOutletName.TEST, strong_action)
        html = render_template()

    assert snapshot == html


def test_outlet_renders_prepended_plugin(snapshot):
    with patch_outlets():
        prepend_template_plugin(PluginOutletName.TEST, strong_action)
        html = render_template()

    assert snapshot == html


def test_outlet_renders_multiple_plugins(snapshot):
    with patch_outlets():
        append_template_plugin(PluginOutletName.TEST, strong_action)
        prepend_template_plugin(PluginOutletName.TEST, em_action)
        prepend_template_plugin(PluginOutletName.TEST, strong_action)
        html = render_template()

    assert snapshot == html


def test_outlet_renders_plugins_with_context(snapshot):
    with patch_outlets():
        append_template_plugin(PluginOutletName.TEST, strong_action)
        prepend_template_plugin(PluginOutletName.TEST, em_action)
        prepend_template_plugin(PluginOutletName.TEST, strong_action)
        html = render_template({"value": "context"})

    assert snapshot == html
