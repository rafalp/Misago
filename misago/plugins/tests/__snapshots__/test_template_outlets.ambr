# serializer version: 1
# name: test_empty_outlet_renders_nothing
  '<div></div>'
# ---
# name: test_outlet_renders_appended_plugin
  '<div><strong>none</strong></div>'
# ---
# name: test_outlet_renders_multiple_plugins
  '<div><strong>none</strong><em>none</em><strong>none</strong></div>'
# ---
# name: test_outlet_renders_plugins_with_context
  '<div><strong>context</strong><em>context</em><strong>context</strong></div>'
# ---
# name: test_outlet_renders_prepended_plugin
  '<div><strong>none</strong></div>'
# ---
